lib/auth/oauth2client.ts

```diff
@@ -0,0 +1,586 @@
+/**
+ * Copyright 2013 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import AuthClient from './authclient';
+import LoginTicket from './loginticket';
+import PemVerifier from './../pemverifier';
+import { merge } from 'lodash';
+import * as querystring from 'querystring';
+import * as util from 'util';
+const noop = Function.prototype;
+
+export default class OAuth2Client extends AuthClient {
+
+  public _clientId: string;
+  public _clientSecret: string;
+  public apiKey: string;
+
+  private _redirectUri: string;
+  private _certificateCache = null;
+  private _certificateExpiry = null;
+  protected _opts: any;
+
+  /**
+   * Handles OAuth2 flow for Google APIs.
+   *
+   * @param {string} clientId The authentication client ID.
+   * @param {string} clientSecret The authentication client secret.
+   * @param {string} redirectUri The URI to redirect to after completing the auth request.
+   * @param {Object} opt_opts optional options for overriding the given parameters.
+   * @constructor
+   */
+  constructor(clientId?: string, clientSecret?: string, redirectUri?: string, opt_opts?: any) {
```

Comment:
Yeah :(  This is just how the code was structured.  ComputeClient (which inherits OAuth2Client) doesn't take any of these properties in it's constructor.  There are also numerous tests that just new up one of these with no parameters.   I suggest we come back around next time and fix this one.